### PR TITLE
Add prettier and pygrep hooks to pre-commit hooks

### PR DESCRIPTION
**Pull request by Armavica**
Towards #8239. This mainly addresses problems in rst and markdown files.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <!-- Feel free to remove check-list items aren't relevant to your change -->
 
-- [ ] Closes #xxxx
-- [ ] Tests added
-- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
-- [ ] New functions/methods are listed in `api.rst`
+- \[ \] Closes #xxxx
```
Comment (max-sixty) [inline .github/PULL_REQUEST_TEMPLATE.md:3]: Are we confident this still displays correctly?

File: design_notes/grouper_objects.md [hunk 2]
```diff
@@ -136,18 +152,22 @@ weighted_sum = xr.dot(ds, weights)
...
 [[1, 0, 0, 1, 1],
  [0, 1, 0, 0, 0],
  [0, 0, 1, 0, 0]]
 ```
-This is the boolean "summarization matrix" referred to in the classic Iverson (1980, Section 4.3)[^2] and "nub sieve" in [various APLs](https://aplwiki.com/wiki/Nub_Sieve).
 
-> [!NOTE]
+This is the boolean "summarization matrix" referred to in the classic Iverson (1980, Section 4.3)\[^2\] and "nub sieve" in [various APLs](https://aplwiki.com/wiki/Nub_Sieve).
+
+> \[!NOTE\]
```
Comment (max-sixty) [inline design_notes/grouper_objects.md:166]: Same here — are we sure these should be escaped? Looks like a directive (though not sure what...)

File: design_notes/grouper_objects.md [hunk 3]
```diff
@@ -227,14 +259,15 @@ Today, it is possible to `ds.groupby(DataArray, ...)`. This syntax will still be
...
 
 ## Copyright
 
 This document has been placed in the public domain.
 
 ## References and footnotes
 
-[^1]:  Wickham, H. (2011). The split-apply-combine strategy for data analysis. https://vita.had.co.nz/papers/plyr.html
-[^2]:  Iverson, K.E. (1980). Notation as a tool of thought. Commun. ACM 23, 8 (Aug. 1980), 444–465. https://doi.org/10.1145/358896.358899
+\[^1\]:  Wickham, H. (2011). The split-apply-combine strategy for data analysis. https://vita.had.co.nz/papers/plyr.html
```
Comment (max-sixty) [inline design_notes/grouper_objects.md:272]: and here; seems unlikely this is correct

File: doc/whats-new.rst
Comment (max-sixty) [inline doc/whats-new.rst:1]: A bunch of changes in this file look quite good!

**Comment by max-sixty** [comment]
Overall looks great!

We should check some of those, I added a couple of comments, though not a complete review.

To the extent that some might interfere with `mypy` or `ruff`, etc, that's not good — we don't want to be in a situation where code can't get through both checkers. Are we OK there?

The markdown & rst checkers look overall great!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (TomNicholas) [reply .github/PULL_REQUEST_TEMPLATE.md:3]: It seems to be [fine](https://github.com/pydata/xarray/blob/097e54955da321b965d00b9fc00d391433680793/.github/PULL_REQUEST_TEMPLATE.md), though I also find that surprising.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -11,6 +11,22 @@ repos:
...
+      - id: python-no-log-warn
+      # - id: python-use-type-annotations  # too many false positives
+      - id: rst-backticks
+      - id: rst-directive-colons
+      - id: rst-inline-touching-normal
+      - id: text-unicode-replacement-char
+  - repo: https://github.com/executablebooks/mdformat
+    rev: 0.7.18
+    hooks:
+    - id: mdformat
```
Comment (keewis) [inline .pre-commit-config.yaml:29]: apparently, we can fix (some of) the surprising bracket escapes with
```suggestion
    - id: mdformat
      additional_dependencies: [mdformat-gfm]
```

Comment (keewis) [reply .pre-commit-config.yaml:29]: though I wonder if instead of using `mdformat` we should rely on `prettier` instead?

Comment (max-sixty) [reply .pre-commit-config.yaml:29]: FWIW in prql we quite successfully use `prettier` to format all non-core code, including yaml / json / md; would recommend: https://github.com/PRQL/prql/blob/86becc591eb1a6421695ccccdf67dd8ab5c775ec/.pre-commit-config.yaml#L18-L22

...and it has a nice VS Code extension (and I guess emacs for the pros among us...)

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (max-sixty) [reply .github/PULL_REQUEST_TEMPLATE.md:3]: Indeed!

Though it seems a bit confusing since people will be directly editing it...

**Comment by headtr1ck** [comment]
Why don't we instead try to rely on ruff and enable these rules there?
Many seems supported already, rules start with PGH*.

**Comment by max-sixty** [comment]
> Why don't we instead try to rely on ruff and enable these rules there? Many seems supported already, rules start with PGH*.

Definitely preferable for the python ones. (though formatting non-python files would still be part of this sort of effort...)

File: CORE_TEAM_GUIDE.md [hunk 5]
```diff
@@ -69,7 +69,7 @@ Here’s a typical workflow for triaging a newly opened issue or discussion:
    The issue tracker is many people’s first interaction with the xarray project itself, beyond just using the library.
    It may also be their first open-source contribution of any kind. As such, we want it to be a welcoming, pleasant experience.
 
-2. **Is the necessary information provided?**
+1. **Is the necessary information provided?**
```
Comment (Illviljan) [inline CORE_TEAM_GUIDE.md:72]: This 1. and the below ones looks wrong.

File: HOW_TO_RELEASE.md [hunk 6]
```diff
@@ -13,109 +13,133 @@ upstream        https://github.com/pydata/xarray (push)
...
-    "ANN xarray").
-    Some of these groups require you to be subscribed in order to email them.
+1. Ensure your main branch is synced to upstream:
+
+   ```sh
+   git switch main
+   git pull upstream main
+   ```
+
+1. Add a list of contributors.
```
Comment (Illviljan) [inline HOW_TO_RELEASE.md:23]: Forgotten the enumeration again.

File: asv_bench/benchmarks/README_CI.md [hunk 7]
```diff
@@ -10,39 +12,39 @@ The `asv` suite can be run for any PR on GitHub Actions (check workflow `.github
...
 
 Due to the sensitivity of the test, we cannot guarantee that false positives are not produced. In practice, values between `(0.7, 1.5)` are to be considered part of the measurement noise. When in doubt, running the benchmark suite one more time will provide more information about the test being a false positive or not.
 
 ## Running the benchmarks on GitHub Actions
 
 1. On a PR, add the label `run-benchmark`.
-2. The CI job will be started. Checks will appear in the usual dashboard panel above the comment box.
-3. If more commits are added, the label checks will be grouped with the last commit checks _before_ you added the label.
-4. Alternatively, you can always go to the `Actions` tab in the repo and [filter for `workflow:Benchmark`](https://github.com/scikit-image/scikit-image/actions?query=workflow%3ABenchmark). Your username will be assigned to the `actor` field, so you can also filter the results with that if you need it.
+1. The CI job will be started. Checks will appear in the usual dashboard panel above the comment box.
```
Comment (Illviljan) [inline asv_bench/benchmarks/README_CI.md:27]: Forgotten enumeration.

File: HOW_TO_RELEASE.md (hunk 6 above)
Comment (dcherian) [reply HOW_TO_RELEASE.md:23]: This is rendering fine. It's a trick to have the renderer do the numbering without having to do it manually

Comment (keewis) [reply HOW_TO_RELEASE.md:23]: the downside of that trick is that reading the source becomes a bit harder to read (which item from the top was that again?), so I'd prefer a linter / autoformatter to keep source and rendered version as close as possible.

Given that [`prettier`](https://github.com/prettier/prettier) ([mirror](https://github.com/rbubley/mirrors-prettier)) appears to get this right, maybe we should be using that instead? That would also auto-format yaml, css and json, and for toml we can try [taplo](https://github.com/tamasfe/taplo/) ([mirror](https://github.com/ComPWA/taplo-pre-commit)).

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (Armavica) [reply .github/PULL_REQUEST_TEMPLATE.md:3]: Ok, then I'll ignore this file.

**Comment by Armavica** [comment]
Thank you for all your comments, I switched to `prettier` which seems to do a less surprising job in many cases.
I also deselected the `pygrep-hooks` that are already checked by `ruff`.

File: doc/whats-new.rst [hunk 8]
```diff
@@ -32,12 +32,30 @@ New Features
...
+  `Tom Nicholas <https://github.com/TomNicholas>`_.
+- A migration guide for users of the prototype `xarray-contrib/datatree repository <https://github.com/xarray-contrib/datatree>`_ has been added, and can be found in the ``DATATREE_MIGRATION_GUIDE.md`` file in the repository root.
+  By `Tom Nicholas <https://github.com/TomNicholas>`_.
+- Added zarr backends for :py:func:`open_groups` (:issue:`9430`, :pull:`9469`).
+  By `Eni Awowale <https://github.com/eni-awowale>`_.
+- Added support for vectorized interpolation using additional interpolators
+  from the ``scipy.interpolate`` module (:issue:`9049`, :pull:`9526`).
+  By `Holly Mandel <https://github.com/hollymandel>`_.
+- Implement handling of complex numbers (netcdf4/h5netcdf) and enums (h5netcdf) (:issue:`9246`, :issue:`3297`, :pull:`9509`).
+  By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
```
Comment (dcherian) [inline doc/whats-new.rst:58]:
```suggestion
```

File: doc/user-guide/dask.rst [hunk 9]
```diff
@@ -381,7 +381,7 @@ In this case, automatic inference has worked so let's check that the result is a
     mapped.identical(ds.time)
 
 Note that we use ``.load(scheduler="single-threaded")`` to execute the computation.
-This executes the Dask graph in `serial` using a for loop, but allows for printing to screen and other
+This executes the Dask graph in ``serial`` using a for loop, but allows for printing to screen and other
```
Comment (dcherian) [inline doc/user-guide/dask.rst:384]:
```suggestion
This executes the Dask graph in serial using a for loop, but allows for printing to screen and other
```

File: doc/internals/internal-design.rst [hunk 10]
```diff
@@ -166,7 +166,7 @@ something interesting:
 
     var._data
 
-You're looking at one of xarray's internal `Lazy Indexing Classes`. These powerful classes are hidden from the user,
+You're looking at one of xarray's internal ``Lazy Indexing Classes``. These powerful classes are hidden from the user,
```
Comment (dcherian) [inline doc/internals/internal-design.rst:169]:
```suggestion
You're looking at one of xarray's internal Lazy Indexing Classes. These powerful classes are hidden from the user,
```

File: doc/internals/how-to-add-new-backend.rst [hunk 11]
```diff
@@ -469,8 +469,8 @@ combining multiple input list with ``itertools.product()``:
 The ``OUTER_1VECTOR`` indexing shall supports number, slices and at most one
 list. The behaviour with the list shall be the same of ``OUTER`` indexing.
 
-If you support more complex indexing as `explicit indexing` or
-`numpy indexing`, you can have a look to the implementation of Zarr backend and Scipy backend,
+If you support more complex indexing as ``explicit indexing`` or
+``numpy indexing``, you can have a look to the implementation of Zarr backend and Scipy backend,
```
Comment (dcherian) [inline doc/internals/how-to-add-new-backend.rst:473]:
```suggestion
If you support more complex indexing as "explicit indexing" or
"numpy indexing", you can have a look to the implementation of Zarr backend and Scipy backend,
```

File: doc/user-guide/data-structures.rst [hunk 12]
```diff
@@ -539,7 +539,7 @@ parent is known as a "root" node (represented by the ``parent`` attribute
 pointing to ``None``). Nodes can have multiple children, but as each child node
 has at most one parent, there can only ever be one root node in a given tree.
 
-The overall structure is technically a `connected acyclic undirected rooted graph`,
+The overall structure is technically a ``connected acyclic undirected rooted graph``,
```
Comment (dcherian) [inline doc/user-guide/data-structures.rst:542]:
```suggestion
The overall structure is technically a connected acyclic undirected rooted graph,
```

There's a lot of these !

File: doc/user-guide/indexing.rst [hunk 13]
```diff
@@ -323,7 +323,7 @@ Vectorized Indexing
 -------------------
 
 Like numpy and pandas, xarray supports indexing many array elements at once in a
-`vectorized` manner.
+``vectorized`` manner.
```
Comment (dcherian) [inline doc/user-guide/indexing.rst:326]:
```suggestion
vectorized manner.
```

File: doc/user-guide/data-structures.rst [hunk 14]
```diff
@@ -608,7 +608,7 @@ We have created a tree with three nodes in it:
 
 
 
-Consistency checks are enforced. For instance, if we try to create a `cycle`,
+Consistency checks are enforced. For instance, if we try to create a ``cycle``,
```
Comment (dcherian) [inline doc/user-guide/data-structures.rst:611]:
```suggestion
Consistency checks are enforced. For instance, if we try to create a *cycle*,
```

File: doc/whats-new.rst (hunk 8 above)
Comment (Armavica) [reply doc/whats-new.rst:58]: Oops, rebase fail… Good catch

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -24,6 +36,10 @@ repos:
       - id: blackdoc
         exclude: "generate_aggregations.py"
         additional_dependencies: ["black==24.8.0"]
+  - repo: https://github.com/rbubley/mirrors-prettier
+    rev: v3.3.3
+    hooks:
+      - id: prettier
```
[the remark this answers is not in the document]
Comment (Armavica) [reply .pre-commit-config.yaml:42]: It looks like it is not supposed to write a cache unless `--cache` is specified: https://prettier.io/docs/en/cli.html#--cache

File: doc/user-guide/weather-climate.rst [hunk 16]
```diff
@@ -148,7 +148,7 @@ use `pandas` when possible, i.e. when the calendar is standard and dates are wit
 The data is unchanged, only the timestamps are modified. Further options are implemented
 for the special ``"360_day"`` calendar and for handling missing dates. There is also
 :py:meth:`xarray.Dataset.interp_calendar` (and :py:meth:`xarray.DataArray.interp_calendar`)
-for `interpolating` data between calendars.
+for ``interpolating`` data between calendars.
```
Comment (dcherian) [inline doc/user-guide/weather-climate.rst:151]:
```suggestion
for interpolating data between calendars.
```

**Pull request merged**